[GTK][WPE] Use a Function for did composite notification

carlosgcampos opened the pull request:
#### 1ce9a2165bf17e4a61b8aff627fc52f5fd69f2d4
<pre>
[GTK][WPE] Use a Function for did composite notification
<a href="https://bugs.webkit.org/show_bug.cgi?id=302783">https://bugs.webkit.org/show_bug.cgi?id=302783</a>

Reviewed by Nikolas Zimmermann.

Rename requestComposition as requestCompositionForRenderingUpdate that
now receives a Function that is only called when the composition for
the rendering update is done. This way we don&apos;t need the composition
request/response previously used to identify the rendering update
composition.

Canonical link: <a href="https://commits.webkit.org/303384@main">https://commits.webkit.org/303384@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f2a70b6e351b368e96b619d318a4e6e420cddb3e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132031 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4523 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43048 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139545 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83939 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4466 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4285 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100930 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68302 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134977 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/3181 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118255 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81720 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/3078 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/951 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82765 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111833 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142192 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4193 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36956 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109300 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4274 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3654 "Found 3 new test failures: http/tests/webgpu/webgpu/api/operation/labels.html http/tests/webgpu/webgpu/api/operation/sampling/lod_clamp.html imported/w3c/web-platform-tests/mediacapture-streams/overconstrained_error.https.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109473 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3178 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114532 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57414 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20549 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4247 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32927 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4078 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67693 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4338 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4206 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
```diff
@@ -124,7 +124,8 @@ LayerTreeHost::~LayerTreeHost()
 #endif
 
-    m_compositor->invalidate();
+    auto compositor = WTFMove(m_compositor);
```
That change seems subtle, why do you need to move it? A comment would be helpful.

carlosgcampos replied:
Because the completion handler should be called, it will be called in the threaded compositor invalidate, but we want to do nothing in that case, so we just check if compositor is nullptr from the callback. I could change it to use a Function, and we don't need to care about ensuring the completion handler is done.

carlosgcampos replied:
I think it would be clearer if we add an Invalid state, so that after invalidate instead of switching to Idle we switch to Invalid, and then we can check the state to return early from compositor.

nikolaszimmermann replied:
Indeed that would be clearer, but maybe use `Destruction` or `Destructing` as phase name?

nikolaszimmermann replied:
Happy to approve, after introducing the new state.

carlosgcampos replied:
In the end switched to use a Function, since we don't really need to ensure the completion handler is called in this case and code is simpler now even without the invalid state.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
```diff
@@ -115,11 +115,15 @@ void ThreadedCompositor::invalidate()
+        completionHandler = std::exchange(m_state.renderingUpdateRequestCompletionHandler, nullptr);
         m_state.state = State::Idle;
     }
+    if (completionHandler)
```
Under what circumstances can this be zero?

carlosgcampos replied:
When there isn't a pending composition for a rendering update. We do this here only because the completion handler should always be called.

nikolaszimmermann replied:
got it.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
```diff
@@ -408,8 +415,13 @@ RunLoop* ThreadedCompositor::runLoop()
+        Locker locker { m_state.lock };
+        completionHandler = std::exchange(m_state.renderingUpdateRequestCompletionHandler, nullptr);
+    }
+    if (completionHandler)
```
Same question, didn't grasp when this can be zero, and why we're in this function if we didn't have a CompletionHandler in the first place?

carlosgcampos replied:
When this is called after invalidate

nikolaszimmermann replied:
Makes sense! I didn't realize those code paths are fired after invalidate, but it can happen.

nikolaszimmermann reviewed (approved):
Much nicer to follow the repaint async code now, too.

carlosgcampos committed:
[GTK][WPE] Use a Function for did composite notification
https://bugs.webkit.org/show_bug.cgi?id=302783

Reviewed by Nikolas Zimmermann.

Rename requestComposition as requestCompositionForRenderingUpdate that
now receives a Function that is only called when the composition for
the rendering update is done. This way we don't need the composition
request/response previously used to identify the rendering update
composition.

Canonical link: https://commits.webkit.org/303384@main
